mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,207 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2025 The Meson development team
+
+from __future__ import annotations
+
+import itertools
+import typing as T
+from pathlib import Path
+
+from .. import build, mesonlib
+from ..interpreter.type_checking import CT_INPUT_KW
+from ..interpreterbase import ContainerTypeInfo, KwargInfo
+from ..interpreterbase.decorators import typed_kwargs, typed_pos_args
+from . import ExtensionModule, ModuleInfo, ModuleReturnValue
+
+if T.TYPE_CHECKING:
+    from typing_extensions import TypedDict
+
+    from ..interpreter import Interpreter
+    from ..programs import ExternalProgram
+    from . import ModuleState
+
+    class ProjectKwargs(TypedDict):
+        sources: T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        include_directories: T.List[T.Union[str, build.IncludeDirs]]
+        link_with: T.List[T.Union[build.SharedLibrary, build.StaticLibrary]]
+        tags: T.List[str]
+        native: bool
+        gomod_tidy: bool
+
+
+class GoBuildModule(ExtensionModule):
+    INFO = ModuleInfo("Go build", "1.6.0", unstable=True)
+    counter: int = 0
+
+    def __init__(self, interpreter: Interpreter) -> None:
+        super().__init__(interpreter)
+        self.go_exe: T.Union[ExternalProgram, build.Executable, None] = None
+        self.methods.update(
+            {
+                "project": self.project,
+            }
+        )
+
+    def detect_tools(self, state: ModuleState, is_native: bool) -> None:
+        self.go_exe = state.find_program("go")
+
+        if not is_native and state.environment.is_cross_build():
+            state.find_program("x86_64-w64-mingw32-gcc")
+            state.find_program("x86_64-w64-mingw32-g++")
+
+    @typed_pos_args(
+        "gobuild.project",
+        str,
+        varargs=(
+            str,
+            mesonlib.File,
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
```

Comment:
There's a shared one of these that does a convesion to MachineChoice you probably want to use, I'll point it out later why.